evm-node.md

```diff
@@ -0,0 +1,338 @@
+           <origStorage> ORIGSTORAGE => ORIGSTORAGE [ INDEX <- #getStorageData(ACCT, INDEX) ] </origStorage>
+           ...
+         </account>
+      requires notBool INDEX in_keys(STORAGE)
+
+    rule <k> #lookupStorage ACCT INDEX => . ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <storage> STORAGE:Map </storage>
+           ...
+         </account>
+      requires INDEX in_keys(STORAGE)
+
+    rule <k> #lookupStorage ACCT _ => . ... </k>
+      requires notBool #accountExists(ACCT)
+```
+
+-   `#getCode` loads the code for a specified account by its address. If the code has already been loaded, it does nothing.
+    If the account does not exist, it also does nothing.
+
+```{.k .node}
+    syntax String ::= #getCode ( Int ) [function, hook(BLOCKCHAIN.getCode)]
+ // -----------------------------------------------------------------------
+    rule <k> #lookupCode ACCT => . ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <code> #unloaded(_) => #parseByteStackRaw(#getCode(ACCT)) </code>
+           ...
+         </account>
+
+    rule <k> #lookupCode ACCT => . ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <code> _:ByteArray </code>
+           ...
+         </account>
+
+    rule <k> #lookupCode ACCT => . ... </k>
+      requires notBool #accountExists(ACCT)
+```
+
+-   `#getBlockhash(N)` returns the blockhash of the Nth most recent block, up to a maximum of 256 blocks.
+    It is used in the implementation of the BLOCKHASH instruction as seen below.
+
+```{.k .node}
+    syntax Int ::= #getBlockhash ( Int ) [function, hook(BLOCKCHAIN.getBlockhash)]
+ // ------------------------------------------------------------------------------
+    rule <k> BLOCKHASH N => #getBlockhash(N) ~> #push ... </k> <mode> NORMAL </mode> requires N >=Int 0 andBool N  <Int 256
+    rule <k> BLOCKHASH N => 0                ~> #push ... </k> <mode> NORMAL </mode> requires N  <Int 0  orBool N >=Int 256
+```
+
+```{.k .node}
+    rule <k> EXTCODEHASH ACCT => HASH ~> #push ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <code> #unloaded(HASH) </code>
+           ...
+         </account>
+```
+
+### Transaction Execution
+
+-   `CREATE2` needs to be modified for the node
+Note that we cannot execute #loadAccount during the #load phase earlier because gas will not yet
+have been paid, and it may be to expensive to compute the hash of the init code.
+
+```{.k .node}
+    rule <k> CREATE2 VALUE MEMSTART MEMWIDTH SALT
+          => #loadAccount #newAddr(ACCT, SALT, #range(LM, MEMSTART, MEMWIDTH))
+          ~> #checkCall ACCT VALUE
+          ~> #create ACCT #newAddr(ACCT, SALT, #range(LM, MEMSTART, MEMWIDTH)) VALUE #range(LM, MEMSTART, MEMWIDTH)
+          ~> #codeDeposit #newAddr(ACCT, SALT, #range(LM, MEMSTART, MEMWIDTH))
+         ...
+         </k>
+         <id> ACCT </id>
```

Comment:
Similar trick with `<loaded>` above, we don't want to override the existing rules for create, just have it inject a `#load` statement right before the CREATE2.